kmk/extensions/oled.py

```diff
@@ -0,0 +1,291 @@
+                                label_direction=view[8],
+                                line_spacing=view[9],
+                                padding_left=1,
+                            )
+                        )
+                    elif view[5] == OledEntryType.IMG:
+                        splash.append(
+                            displayio.TileGrid(
+                                view[0],
+                                pixel_shader=view[0].pixel_shader,
+                                x=view[1],
+                                y=view[2],
+                            )
+                        )
+        gc.collect()
+        self._display.show(splash)
+
+    def updateOLED(self, sandbox):
+        self.render_oled(sandbox.active_layers[0])
+        gc.collect()
+
+    def on_runtime_enable(self, sandbox):
+        return
+
+    def on_runtime_disable(self, sandbox):
+        return
+
+    def during_bootup(self, keyboard):
+        if self._split is not None:
+            if self._split.split_side == 1:
+                self.split_side = 'L'
+                self._flip = self._flip_left
+            elif self._split.split_side == 2:
+                self.split_side = 'R'
+                self._flip = self._flip_right
+        else:
+            self.split_side = None
+
+        for idx, view in enumerate(self._views):
+            if view[3] != self.split_side and view[3] is not None:
+                del self._views[idx]
+
+        displayio.release_displays()
+        i2c = busio.I2C(keyboard.SCL, keyboard.SDA)
+        self._display = adafruit_displayio_ssd1306.SSD1306(
+            displayio.I2CDisplay(i2c, device_address=self._device_address),
+            width=self._width,
+            height=self._height,
+            rotation=180 if self._flip else 0,
+            brightness=self._brightness,
+        )
+        self.render_oled(0)
+        return
+
+    def before_matrix_scan(self, sandbox):
+        self.dim()
+        if self._dark != self._go_dark or sandbox.active_layers[0] != self._prevLayers:
+            self._dark = self._go_dark
+            self._prevLayers = sandbox.active_layers[0]
+            self.updateOLED(sandbox)
+        return
+
+    def after_matrix_scan(self, keyboard):
+        if keyboard.matrix_update or keyboard.secondary_matrix_update:
+            self.timer_time_reset()
+        return
+
+    def before_hid_send(self, sandbox):
+        return
+
+    def after_hid_send(self, sandbox):
+        return
+
+    def on_powersave_enable(self, sandbox):
+        self._powersave = True
+        return
+
+    def on_powersave_disable(self, sandbox):
+        self._powersave = False
+        return
+
+    def _oled_bri(self, *args, **kwargs):
+        self._display.brightness = (
+            self._display.brightness + self._brightness_step
+            if self._display.brightness + self._brightness_step <= 1.0
+            else 1.0
+        )
```

Comment:
There's a `clamp` function in `kmk.utils` for that.
```suggestion
        self._display.brightness = clamp(self._display.brightness + self._brightness_step, 0, 1)
```

Reply:
Changed that. 👍